FetchDlang/fetchdlang.d

```diff
@@ -0,0 +1,475 @@
+
+enum ResultType
+{
+    failure,
+    success,
+}
+
+enum helpText =
+    "
+
+    https://github.com/D-Programming-Language/tools/blob/master/update.sh
+    
+    1. create a working directory where the source files are to be installed
+    2. change to this directory
+    3. run fetchdlang from this directory.
+       eg /usr/sbin/fetchdlang when you are in /opt/dlangdev
+    
+
+    The following subdirectories will be created, each containing the latest
+    sources from github:
+        dmd,druntime,phobos,dlang.org,tools,installer
+    
+    After source has been fetched, the tool will build each component in turn.
+
+    On on ongoing basis, to update your toolchain from github, go to the same
+    directory and run fetchdlang again.  fetchdlang will detect that directories
+    exist and will do an update.";
+
+enum argumentHelpText = 
+"    Arguments
+    ---------
+    No arguments are required, but the following are options:
+
+    install:      copy the D compiler from the dmd build directory and replace
+                  local binary install
+                  *** not advisable *** unless you know what you are doing, as
+                  you may break your local DMD
+
+    onlyinstall:  do not fetch or build anything - locally install dmd binaries
+                  from build directory
+
+    --tag=XYZ:    fetch, build, and (optionallly install) a tagged version from
+                  the repositories.
+                  Eg --tag = 2.067.0-b1 for DMD beta 2.067 beta 1.  This will
+                  prepend the tag argumeny by 'v' and fetch v2.067-0-b1 from git.
+
+    --cleanlocks: Only use this if you are sure you know what you are doing. 
+                  Incomplete git fetch may lock some of the repositories.  In
+                  this case, to update them you will need to remove the
+                  .git/index.lock file that it leaves behind.  This option
+                  removes the lock, but can cause problems if the reason for the
+                  lock is that a concurrent instance of git is running.";
+
+
+debug
+{
+    void log(string msg)
+    {
+        stderr.writefln(msg);
+    }
+}
+
+struct Params
+{
+    bool cleanLocks = false;
+    string tag;
+    bool installDMD = false;
+    bool onlyInstall = false;
+    string workingDir;
+    string tempDir;
+    string[] toUpdate;
+    string[] toInstall;
+}
+
+int main(string[] args)
+{
+    import std.path : dirSeparator;
+    import scriptutil : makeTempDir;
+    import std.string : indexOf;
+    import std.file: getcwd, tempDir, mkdirRecurse, rmdirRecurse;
+
+    Params params;
+    params.tempDir = makeTempDir(tempDir() ~ dirSeparator ~ "dmd-update.XXX"); // working directory
+    params.workingDir = getcwd();
+
+// Take care of the command line arguments
+    writefln("* fetchdlang: a tool for building and maintaining the dlang toolchain from github");
+    foreach(arg; args[1 .. $])
```

Comment:
honestly, as I wrote it, I sort of clenched my teeth and forced myself to choose the path of a line of line translation rather than a  write the functionality from scratch path.  if I were writing from scratch, I would also have made as much as possible data rather than hard coding it - so you have a json table of what you need to do.  but it is harder to understand, and harder to see what it is doing that way.

On the one hand, the reason to have done this is because it is appealing to have a native code tool and to eat your own dogfood.  On the other everything that goes in here is somewhat a showcase, so it makes sense to do things as elegantly as possible.

I am happy to change if people feel strongly about it.

Wrt digger vs a simple little tool, I wasn't aware of digger, and it is beautiful, like Vladimir's other projects.  It is a philosophical choice about simple tools that do one thing versus powerful tools that unavoidably are more complex and intimidating.  I do not think it hurts to give people a choice, because of the compounding of complexity problem when you are new to a domain.  I think thtat's why Andrei was excited about update.sh - his focus being to make it easy and unintimidating for new developers to get involved, and because small initial frictions have large consequences.

Anyway - I leave it to you to decide.


Laeeth.